cuda/entry_state.hpp

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+#include <optional>
+#include <type_traits>
+#include <unordered_map>
+#include <vector>
+
+#include "xprof_utils.hpp"
+
+
+class EntryState {
+public:
+  template <class K,
+            typename = std::enable_if_t<std::is_trivially_copyable_v<K>>>
+  void push_entry(hpt_t hpt, K v) {
+    const auto *b = (std::byte *)&v;
+    std::vector<std::byte> res{b, b + sizeof(K)};
+    push_entry_impl(hpt, res);
+  }
+
+  template <class K,
+            typename = std::enable_if_t<std::is_trivially_copyable_v<K>>>
+  K pop_entry(hpt_t hpt) {
+    auto &v = pop_entry_impl(hpt);
+    const K res{*(K *)(v.value().data())};
+    v.reset();
+    return res;
+  }
+
+  //-- String specialization
+  void push_entry(hpt_t hpt, const std::string s) {
+    const auto *b = (std::byte *)s.data();
+    std::vector<std::byte> res{b, b + s.size() + 1};
+    push_entry_impl(hpt, res);
+  }
+
+  template <class K,
+            typename = std::enable_if_t<std::is_same_v<K, std::string>>>
+  std::enable_if_t<std::is_same_v<K, std::string>, K>
+  pop_entry(hpt_t hpt) {
+    auto &v = pop_entry_impl(hpt);
+    const std::string res{(char *)v.value().data()};
+    v.reset();
+    return res;
+  }
+
+  void set_ts(hpt_t hpt, int64_t ts) {
```

Comment:
I guess we should choose between `pop / push` and `set / get`. They are doing the same thing.

I vote maybe more for `set / get` , as we cannot push multiple entry for the `hpt`, it will be overwritten.

Reply:
ya I like set / get. I wonder about the cost of doing .reset also, I guess it is just deleting a key not zeroing values so could be quite fast.

Reply:
So let rename `push/ pop`  `set / get` then :)